### ISSUE TITLE
How to calculate the active weight for GSC?

### ISSUE THREAD

**Issue by zacharyLEE26** [kkumatani/distant_speech_recognition#2]
Hi,
Thank you for updating BTK 2.0! I had been used 1.0 for a long time, and I was glad to see the toolkit was still being updated.
In the previous version, I could not found how to calculate the active weights for sidelobe canceller, as if the packedweights should be provided previously. Do you have some idea for updating the adaptive weights? I had tried DAS and SD beamformer, and the fixed beamforming worked well, however, neither non-adaptive nor adaptive ways I tried didn't work. Could you help me?
Thank you!

**Comment by kkumatani** [comment]
Yes, I plan to add that example to https://distantspeechrecognition.sourceforge.io/btk20_documentation/user_docs/beamforming.html#

Stay tuned.

**Comment by zacharyLEE26** [comment]
Hi,
Thanks for your reply! I had set up BTK2.0 and worked correctly.

Recently, I stepped into test_online_beamforming.py for testing GSCLMSBeamformer , when I stepped into the python file, pybeamformer.py , I was confused with some snippet code.

After getting through Griffiths Jim's reference, he said that before the signal was implemented to the adaptive algorithm,  the signal should be calculated with time-delay steering elements so as to point in the direction of interest （some other reference might also call it time alignment）.  However, I didn't find such implementation (approximately at lines 614 in pybeamformer.py). I'm afraid that is there any wrong with my thought?

I hope that  you could help me to solve my confusion, thank you very much!

**Comment by zacharyLEE26** [comment]
Hi,
Thanks for your reply! I had set up BTK2.0 and worked correctly.
Recently, I stepped into test_online_beamforming.py for testing GSCLMSBeamformer , when I stepped into the python file, pybeamformer.py , I was confused with some snippet code.
After getting through Griffiths Jim's reference, he said that before the signal was implemented to the adaptive algorithm, the signal should be calculated with time-delay steering elements so as to point in the direction of interest （some other reference might also call it time alignment）. However, I didn't find such implementation (approximately at lines 614 in pybeamformer.py). I'm afraid that is there any wrong with my thought?
I hope that you could help me to solve my confusion, thank you very much!

zili zhou

>

**Comment by kkumatani** [comment]
Hi ZacharyLEE26,

Your question is perhaps where the time-delay steering vector computation or time alignment has been done in GSCLMSBeamformer.

Let us follow Harry L. Van Trees book's terminology here; see section 6.7.3. The GSC beamformer consists of the three blocks:
1. Quiescent weight vector at an upper branch **wq**, 
2. Blocking matrix **B** and 
3. Active weight vector **wa** at a lower branch.
Note that the weight vector/matrix are computed per frequency bin. 

The quiescent weight vector corresponds to the delay-and-sum beamformer's weight vector. The quiescent vector is calculated with time-delay steering elements so as to point in the direction of interest. That is done [at line 694 of pybeamformer.py](https://github.com/kkumatani/distant_speech_recognition/blob/master/btk20_src/lib/pybeamformer.py#L694)
Notice that the function takes the time delay for the direction of interest. 

The inner product operation of the conjugated quiescent weight vector and input vector is referred to as time alignment. It is also done at  [at line 632 of pybeamformer.py](
https://github.com/kkumatani/distant_speech_recognition/blob/master/btk20_src/lib/pybeamformer.py#L632)
It will literally align the time delay of the planar wavefront.

That was at Python level. By the way, BTK also provides the C++ implementation for that; see https://github.com/kkumatani/distant_speech_recognition/blob/master/btk20_src/beamformer/beamformer.h#L181

**Comment by kkumatani** [comment]
I see what you mean. 

In the BTK, **B** is computed with Gram Schmidt orthogonalization so as to satisfy **wq**^H **B** = **0** and **B** ^H **B** = **I**. Accordingly, it does not require phase pre-steering before the upper and lower branches (time alignment). In Griffiths and Jim's paper, the quiescent vector and blocking matrix are fixed for no time delay case. That is why it needs pre-steering.
